src/detect-bytemath.c

```diff
@@ -0,0 +1,1196 @@
+                       "value: \"%s\"", tmp_str);
+            goto error;
+        }
+
+        /* determine how many trailing 0's are in the bitmask. This will be used
+         * to rshift the value after applying the bitmask
+         */
+        bmd->bitmask_shift_count = 0;
+        if (bmd->bitmask_val) {
+            uint32_t bmask = bmd->bitmask_val;
+            while (!(bmask & 0x1)){
+                bmask = bmask >> 1;
+                bmd->bitmask_shift_count++;
+            }
+        }
+    }
+
+    if (bmd->endian == DETECT_BYTEMATH_ENDIAN_DCE) {
+        switch (bmd->nbytes) {
+            case 1:
+            case 2:
+            case 4:
+                break;
+            default:
+                SCLogError(SC_ERR_INVALID_SIGNATURE, "nbytes must be 1, 2, or 4 "
+                           "when used with \"dce\"; %d is not valid", bmd->nbytes);
+                goto error;
+                break;
+        }
+    }
+
+    switch (bmd->oper) {
+        case DETECT_BYTEMATH_OPERATOR_LSHIFT:
+        case DETECT_BYTEMATH_OPERATOR_RSHIFT:
+            /* nbytes has already been validated to be in the range [1, 10] */
+            if (bmd->nbytes > 4) {
+                SCLogError(SC_ERR_INVALID_SIGNATURE, "nbytes must be 1 through 4 (inclusive) "
+                           "when used with \"<<\" or \">>\"; %d is not valid", bmd->nbytes);
+                goto error;
+            }
+            break;
+
+        default:
+            break;
+    }
+
+    /* Set defaults for endian and base if needed */
+    if (!(bmd->flags & DETECT_BYTEMATH_FLAG_ENDIAN)) {
+        bmd->endian = DETECT_BYTEMATH_ENDIAN_DEFAULT;
+    }
+    if (!(bmd->flags & DETECT_BYTEMATH_FLAG_STRING)) {
+        bmd->base = DETECT_BYTEMATH_BASE_DEFAULT;
+    }
+
+    return bmd;
+
+ error:
+    if (bmd != NULL)
+        DetectByteMathFree(de_ctx, bmd);
+    return NULL;
+}
+
+/**
+ * \brief The setup function for the byte_math keyword for a signature.
+ *
+ * \param de_ctx Pointer to the detection engine context.
+ * \param s      Pointer to signature for the current Signature being parsed
+ *               from the rules.
+ * \param arg    Pointer to the string holding the keyword value.
+ *
+ * \retval  0 On success.
+ * \retval -1 On failure.
+ */
+static int DetectByteMathSetup(DetectEngineCtx *de_ctx, Signature *s, const char *arg)
+{
+    SigMatch *sm = NULL;
+    SigMatch *prev_pm = NULL;
+    DetectByteMathData *data;
+    char *rvalue = NULL;
+    int ret = -1;
+
+    data = DetectByteMathParse(de_ctx, arg, &rvalue);
+    if (data == NULL)
+        goto error;
+
+    int sm_list;
+    if (s->init_data->list != DETECT_SM_LIST_NOTSET) {
+        if (DetectBufferGetActiveList(de_ctx, s) == -1)
+            goto error;
+
+        sm_list = s->init_data->list;
+
+        if (data->flags & DETECT_BYTEMATH_FLAG_RELATIVE) {
+            prev_pm = DetectGetLastSMFromLists(s, DETECT_CONTENT, DETECT_PCRE, -1);
```

Comment:
should we check the result here?

Reply:
Good question. `prev_pm` is null-checked later.

Reply:
Are you referring to the check at line 683? That would suggest its okay to have a byte_math setting of 'relative', while there is no previous match to be relative to.

Reply:
Yes -- that's the check I'm referring to. 

I'll have to modify the other `byte-` methods to return an error in this case as they use this pattern. I'll update my pr.

Reply:
In the other keywords it may be deliberate, not sure. May be for compatibility reasons with older rules. Guess we can test with the ET set.